File: modules/pdf-renderer/android/src/main/java/expo/modules/pdfrenderer/PdfRendererModule.kt
package expo.modules.pdfrenderer

import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.pdf.PdfRenderer
import android.net.Uri
import android.os.ParcelFileDescriptor
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream

/**
 * Rasterises PDF pages with the platform renderer so the editor can show the
 * user the actual document they are signing.
 *
 * android.graphics.pdf.PdfRenderer reports page size in points at 72 dpi, which
 * is the same coordinate space pdf-lib writes into, so a tap mapped through the
 * render ratio lands where the export puts it.
 */
class PdfRendererModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("PdfRenderer")

    AsyncFunction("getInfo") { uri: String, promise: Promise ->
      try {
        withRenderer(uri) { renderer ->
          val pages = (0 until renderer.pageCount).map { index ->
            renderer.openPage(index).use { page ->
              mapOf(
                "pointWidth" to page.width.toDouble(),
                "pointHeight" to page.height.toDouble(),
              )
            }
          }
          promise.resolve(mapOf("pageCount" to renderer.pageCount, "pages" to pages))
        }
      } catch (error: Exception) {
        promise.reject("ERR_PDF", error.message ?: "The PDF could not be opened.", error)
      }
    }

    AsyncFunction("renderPage") {
      uri: String, pageIndex: Int, targetWidth: Double, promise: Promise ->
      try {
        withRenderer(uri) { renderer ->
          if (pageIndex < 0 || pageIndex >= renderer.pageCount) {
            promise.reject("ERR_PDF", "Page $pageIndex is out of range.", null)
            return@withRenderer
          }

          renderer.openPage(pageIndex).use { page ->
            val scale = targetWidth / page.width
            val pixelWidth = Math.round(page.width * scale).toInt().coerceAtLeast(1)
            val pixelHeight = Math.round(page.height * scale).toInt().coerceAtLeast(1)

            val bitmap = Bitmap.createBitmap(pixelWidth, pixelHeight, Bitmap.Config.ARGB_8888)
            // PdfRenderer composites onto whatever is already there, so an
            // unfilled bitmap leaves transparent gaps where the page is blank.
            bitmap.eraseColor(Color.WHITE)
            page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)

            val destination = File(
              appContext.cacheDirectory,
              "signpure_page_${pageIndex}_${System.currentTimeMillis()}.jpg"
            )
            FileOutputStream(destination).use { out ->
              bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out)
            }
            bitmap.recycle()

            promise.resolve(
              mapOf(
                "uri" to Uri.fromFile(destination).toString(),
                "pixelWidth" to pixelWidth,
                "pixelHeight" to pixelHeight,
                "pointWidth" to page.width.toDouble(),
                "pointHeight" to page.height.toDouble(),
              )
            )
          }
        }
      } catch (error: Exception) {
        promise.reject("ERR_PDF", error.message ?: "The page could not be rendered.", error)
      }
    }
  }

  private inline fun withRenderer(uri: String, block: (PdfRenderer) -> Unit) {
    val context = appContext.reactContext ?: error("No React context available.")
    val parsed = Uri.parse(uri)
    val descriptor: ParcelFileDescriptor = if (parsed.scheme == "content") {
      context.contentResolver.openFileDescriptor(parsed, "r")
        ?: error("The document could not be opened.")
    } else {
      ParcelFileDescriptor.open(
        File(parsed.path ?: uri),
        ParcelFileDescriptor.MODE_READ_ONLY
      )
    }
    descriptor.use { fd -> PdfRenderer(fd).use(block) }
  }
}

File: modules/pdf-renderer/ios/PdfRendererModule.swift
import ExpoModulesCore
import PDFKit
import UIKit

/// Rasterises PDF pages with PDFKit so the editor can show the user the actual
/// document they are signing.
///
/// The previous editor drew a generic mock-up of a contract and placed
/// signatures onto that, so the user never saw their own file and the
/// coordinates bore no relation to it.
public class PdfRendererModule: Module {
  public func definition() -> ModuleDefinition {
    Name("PdfRenderer")

    AsyncFunction("getInfo") { (uri: String, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        guard let document = Self.open(uri) else {
          promise.reject("ERR_PDF", "The PDF could not be opened.")
          return
        }

        var pages: [[String: Any]] = []
        for index in 0..<document.pageCount {
          guard let page = document.page(at: index) else { continue }
          let bounds = page.bounds(for: .mediaBox)
          pages.append([
            "pointWidth": bounds.width,
            "pointHeight": bounds.height,
          ])
        }

        promise.resolve(["pageCount": document.pageCount, "pages": pages])
      }
    }

    AsyncFunction("renderPage") {
      (uri: String, pageIndex: Int, targetWidth: Double, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        guard let document = Self.open(uri), let page = document.page(at: pageIndex) else {
          promise.reject("ERR_PDF", "Page \(pageIndex) could not be read.")
          return
        }

        // mediaBox is the same box pdf-lib measures against, so a tap mapped
        // through this ratio lands where the export writes.
        let bounds = page.bounds(for: .mediaBox)
        guard bounds.width > 0, bounds.height > 0 else {
          promise.reject("ERR_PDF", "Page \(pageIndex) has no usable media box.")
          return
        }

        let scale = targetWidth / bounds.width
        let pixelSize = CGSize(
          width: (bounds.width * scale).rounded(),
          height: (bounds.height * scale).rounded()
        )

        let format = UIGraphicsImageRendererFormat.default()
        format.scale = 1  // pixelSize is already in pixels.
        format.opaque = true

        let image = UIGraphicsImageRenderer(size: pixelSize, format: format).image { context in
          UIColor.white.set()
          context.fill(CGRect(origin: .zero, size: pixelSize))

          // PDF space has its origin bottom-left; flip into image space.
          context.cgContext.translateBy(x: 0, y: pixelSize.height)
          context.cgContext.scaleBy(x: scale, y: -scale)
          context.cgContext.translateBy(x: -bounds.origin.x, y: -bounds.origin.y)
          page.draw(with: .mediaBox, to: context.cgContext)
        }

        guard let data = image.jpegData(compressionQuality: 0.9) else {
          promise.reject("ERR_PDF", "The rendered page could not be encoded.")
          return
        }

        let destination = FileManager.default.temporaryDirectory
          .appendingPathComponent("signpure_page_\(pageIndex)_\(Int(Date().timeIntervalSince1970 * 1000)).jpg")

        do {
          try data.write(to: destination)
        } catch {
          promise.reject("ERR_PDF", "The rendered page could not be written: \(error.localizedDescription)")
          return
        }

        promise.resolve([
          "uri": destination.absoluteString,
          "pixelWidth": Int(pixelSize.width),
          "pixelHeight": Int(pixelSize.height),
          "pointWidth": bounds.width,
          "pointHeight": bounds.height,
        ])
      }
    }
  }

  private static func open(_ uri: String) -> PDFDocument? {
    let url = URL(string: uri) ?? URL(fileURLWithPath: uri)
    return PDFDocument(url: url)
  }
}
